### ISSUE TITLE
twisted.trial TestCase should support assertDictEqual from Python 2.7 unittest

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5291]
|<img alt="iuridiniz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| iuridiniz reported|
|-|-|
|Trac ID|trac#5291|
|Type|enhancement|
|Created|2011-10-02 15:34:37Z|
|Branch|https://github.com/twisted/twisted/tree/assertdictequal-5291|

python 2.7 unitest.TestCase.assertDictEqual is not compatible with trial failUnlessIsInstance do not accept msg argument

Versions affected: 11.0.0, 10.2.0, 10.1.0

See the attached files

```
$ trial python_bug_unittest.py
```

```
python_bug_unittest
  TestCase
    testAssertDictEqual ...                                                [OK]
    testAssertIsInstance ...                                               [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 0.023s

PASSED (successes=2)
(test_twisted)iuri@moeda:~/Sources/DIGINET/bug$ trial python_bug_unittest.py 
python_bug_unittest
  TestCase
    testAssertDictEqual ...                                                [OK]
    testAssertIsInstance ...                                               [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 0.021s

PASSED (successes=2)

```

```
$ trial twisted_bug_unittest.py
```

```
twisted_bug_unittest
  TestCase
    testAssertDictEqual ...                                             [ERROR]
    testAssertIsInstance ...                                            [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/iuri/Sources/DIGINET/bug/twisted_bug_unittest.py", line 11, in testAssertDictEqual
    self.assertDictEqual(dict(), dict())
  File "/usr/lib/python2.7/unittest/case.py", line 816, in assertDictEqual
    self.assertIsInstance(d1, dict, 'First argument is not a dictionary')
exceptions.TypeError: failUnlessIsInstance() takes exactly 3 arguments (4 given)

twisted_bug_unittest.TestCase.testAssertDictEqual
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/iuri/Sources/DIGINET/bug/twisted_bug_unittest.py", line 8, in testAssertIsInstance
    self.assertIsInstance(dict(), dict, '')
exceptions.TypeError: failUnlessIsInstance() takes exactly 3 arguments (4 given)

twisted_bug_unittest.TestCase.testAssertIsInstance
-------------------------------------------------------------------------------
Ran 2 tests in 0.015s

FAILED (errors=2)

```

Attachments:

* [python_bug_unittest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/963/96307ff7902e413b1389ccecaf960d7286c3597e/969aa6f48588ab2a48c63445be55285c24fd3b0c.py) (312 bytes) - added by iuridiniz on 2011-10-02 15:35:03Z - 
* [twisted_bug_unittest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/963/96307ff7902e413b1389ccecaf960d7286c3597e/eff2e6683e080b70b0b01f83e11340f1110b25e3.py) (332 bytes) - added by iuridiniz on 2011-10-02 15:35:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5291 5291
type__enhancement enhancement
reporter__iuridiniz iuridiniz
priority__low low
milestone__ 
branch__branches_assertdictequal_5291 branches/assertdictequal-5291
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__trial_assertDictEqual_easy trial assertDictEqual easy
time__1317569677000000 1317569677000000
changetime__1318879621000000 1318879621000000
version__None None
owner__exarkun exarkun
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Thanks.  I guess we should support this, since it's surprising (to end users, if not maintainers) for trial's `TestCase` subclass to not support functionality of the base class.  trial is also more useful the more compatible it is with `unittest.TestCase`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [32892]) Branching to 'assertdictequal-5291'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assertdictequal-5291)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

[1]
```
+            self.fail("%r is not an instance of %s%s" % (
+                    instance, classOrTuple, suffix))
```

The indentation of the second line is slightly off.

+1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

> The indentation of the second line is slightly off.

You almost got me.  emacs likes it there, though, and I guess I do too - two open parens leading up to it merits 8 spaces of indentation from the statement's level, and that's just where it is.  I guess it just comes close enough to being open-paren aligned to _look_ like it should be indented a couple more spaces.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [32905]) Merge assertdictequal-5291

Author: exarkun
Reviewer: therve
Fixes: [#5291](https://github.com/twisted/twisted/issues/5291)

Add a message parameter to `failUnlessIsInstance` (aka `assertIsInstance`).
This lets application code augment the exception message when the assertion
fails.  It also makes the signature compatible with the stdlib unitttest
version, which lets other stdlib unittest code which _uses_
`assertIsInstance` work properly - primarily `assertDictEqual`.
